### PR TITLE
Fix BART attention fusion for SDPA pattern from transformers >= 4.49

### LINKED ISSUE #23864: Attention fusion broken for BART 🤖
### Describe the issue

Thanks for your awesome work on `onnxruntime` :100:.

I noticed that attention fusion is currently broken for BART, when models are exported using a recent Transformers due changes in the SDPA implementation (see [here](https://github.com/huggingface/transformers/commit/2c47618c1a282f925446506d53108dc6e82d9ef0)). @tianleiwu recently updated attention fusion for BERT among others (see #22629). This issue is probably a follow up for BART, which is still broken.

### To reproduce

dependencies:
```
onnx==1.17.0
onnxruntime==1.22.0.dev20250227005
transformers==4.49.0
```

## BART

For BART attention fusion fails:

```python
import torch
from onnxruntime.transformers import optimizer
from onnxruntime.transformers.fusion_options import FusionOptions
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

model_name = "hf-internal-testing/tiny-random-bert"
model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)

model.eval()


class EncoderWrapper(torch.nn.Module):
    def __init__(self, encoder: torch.nn.Module):
        super().__init__()
        self.encoder = encoder

    def forward(
        self, input_ids: torch.Tensor, attention_mask: torch.Tensor
    ) -> torch.Tensor:
        outs = self.encoder(input_ids, attention_mask)
        return outs["last_hidden_state"]


model = EncoderWrapper(encoder=model.model.encoder)
print(model)

text = "Hello, how are you?"
inputs = tokenizer(text, return_tensors="pt")

input_ids = inputs["input_ids"]
attention_mask = inputs["attention_mask"]

input_names = ["input_ids", "attention_mask"]
output_names = ["encoder_output"]

onnx_path = "bart_model.onnx"

torch.onnx.export(
    model,
    (input_ids, attention_mask),
    onnx_path,
    export_params=True,
    input_names=input_names,
    output_names=output_names,
    dynamic_axes={
        "input_ids": {0: "batch_size", 1: "sequence_length"},
        "attention_mask": {0: "batch_size", 1: "sequence_length"},
        "encoder_output": {0: "batch_size", 1: "sequence_length"},
    },
    opset_version=16,
)
print(f"BART encoder exported to {onnx_path}")

optimization_options = FusionOptions("bart")
optimization_options.enable_attention = True


m = optimizer.optimize_model(
    onnx_path,
    model_type="bart",
    num_heads=0,
    hidden_size=0,
    opt_level=2,
    use_gpu=False,
    verbose=True,
    optimization_options=optimization_options,
    only_onnxruntime=False,
)

optimized_path = "bart_encoder_optimized.onnx"
m.save_model_to_file(optimized_path)

print(f"Optimized ONNX model saved to {optimized_path}")
print(m.get_fused_operator_statistics())

```

```
{'EmbedLayerNormalization': 0, 'Attention': 0, 'MultiHeadAttention': 0, 'Gelu': 0, 'FastGelu': 0, 'BiasGelu': 2, 'GemmFastGelu': 0, 'LayerNormalization': 1, 'SimplifiedLayerNormalization': 0, 'SkipLayerNormalization': 4, 'SkipSimplifiedLayerNormalization': 0, 'RotaryEmbedding': 0, 'QOrderedAttention': 0, 'QOrderedGelu': 0, 'QOrderedLayerNormalization': 0, 'QOrderedMatMul': 0}
```

## BERT

For BERT attention fusion is successful.

```python
import torch
from onnxruntime.transformers import optimizer
from transformers import AutoTokenizer, BertModel

model = BertModel.from_pretrained("hf-internal-testing/tiny-random-bert")
tokenizer = AutoTokenizer.from_pretrained("hf-internal-testing/tiny-random-bert")

model.eval()

text = "Hello, how are you?"
inputs = tokenizer(text, return_tensors="pt")
input_names = ["input_ids", "attention_mask"]
output_names = ["output"]

onnx_path = "bert_model.onnx"

torch.onnx.export(
    model,
    (inputs["input_ids"], inputs["attention_mask"]),
    onnx_path,
    input_names=input_names,
    output_names=output_names,
    dynamic_axes={
        "input_ids": {0: "batch_size", 1: "sequence_length"},
        "attention_mask": {0: "batch_size", 1: "sequence_length"},
        "output": {0: "batch_size", 1: "sequence_length"},
    },
    opset_version=20,
    verbose=True,
)
print(f"Model exported to {onnx_path}")

m = optimizer.optimize_model(
    onnx_path,
    model_type="bert",
    num_heads=0,
    hidden_size=0,
    opt_level=2,
    use_gpu="cpu",
    verbose=True,
)

m.save_model_to_file(
    "bert_model_optimized.onnx",
)
print(m.get_fused_operator_statistics())
```

```
{'EmbedLayerNormalization': 1, 'Attention': 5, 'MultiHeadAttention': 0, 'Gelu': 0, 'FastGelu': 0, 'BiasGelu': 5, 'GemmFastGelu': 0, 'LayerNormalization': 0, 'SimplifiedLayerNormalization': 0, 'SkipLayerNormalization': 10, 'SkipSimplifiedLayerNormalization': 0, 'RotaryEmbedding': 0, 'QOrderedAttention': 0, 'QOrderedGelu': 0, 'QOrderedLayerNormalization': 0, 'QOrderedMatMul': 0}
```

**output:**

```
{'EmbedLayerNormalization': 1, 'Attention': 5, 'MultiHeadAttention': 0, 'Gelu': 0, 'FastGelu': 0, 'BiasGelu': 5, 'GemmFastGelu': 0, 'LayerNormalization': 0, 'SimplifiedLayerNormalization': 0, 'SkipLayerNormalization': 10, 'SkipSimplifiedLayerNormalization': 0, 'RotaryEmbedding': 0, 'QOrderedAttention': 0, 'QOrderedGelu': 0, 'QOrderedLayerNormalization': 0, 'QOrderedMatMul': 0}
```

1. Are there any plans to make attention fusion work again for BART like for BERT when exported with a recent transformers version?
2. If I have the option to export using the dynamo backend `torch.export(..., dynamo=True)`, I could try to use onnxscript to perform attention fusion using `onnxscript.rewriter` (see [here.](https://onnxscript.ai/tutorial/rewriter/rewrite_patterns.html)). Is this the preferred/modern way to perform attention fusion?

Happy to contribute a pr. I just have little knowledge about the pattern matching mechanisms in `onnxruntime/python/tools/transformers/fusion_bart_attention.py`, but would be willing to learn.

### Urgency

Medium.

### Platform

Linux

### OS Version

Ubuntu 22.04.5 LTS

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

onnxruntime==1.22.0.dev20250227005

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

_No response_

### PR DESCRIPTION
**Pull request by Rishi-Dave**
## Summary
- Add SDPA-aware pattern matching to `FusionBartAttention` so that BART attention fusion succeeds on models exported with HuggingFace Transformers >= 4.49
- Add a synthetic BART SDPA graph generator and unit test

## Motivation
Fixes #23864

HuggingFace Transformers >= 4.49 replaced `BartAttention` with `BartSdpaAttention` ([commit `2c47618`](https://github.com/huggingface/transformers/commit/2c47618)), changing the ONNX export graph topology in several ways that broke `FusionBartAttention` pattern matching. Running `optimize_model(..., model_type="bart")` on these newer exports produces **zero** fused Attention nodes.

### `fusion_bart_attention.py`

The SDPA refactor introduces four structural changes to the attention subgraph. Each required a new match path:

1. **QKV output path — LayerNormalization anchor fallback**
   For SDPA models, symbolic shape inference often fails, which prevents SkipLayerNormalization fusion. When the anchor node is a plain `LayerNormalization` instead of `SkipLayerNormalization`, there's an extra residual `Add` between the LayerNorm and the attention output projection. Added a fallback match: `["Add", "Add", "MatMul", "Reshape", "Transpose", "MatMul"]` with `[0, None, 0, 0, 0, 0]`.

2. **QK path — NaN guard (Where + IsNaN)**
   SDPA wraps the Softmax output in a NaN guard: `Where(IsNaN(softmax), 0.0, softmax)`. The `Where` node's input[2] is the Softmax output. Added two new QK paths:
   - No mask: `["Where", "Softmax", "MatMul"]` with `[0, 2, 0]`
   - With mask: `["Where", "Softmax", "Add", "MatMul"]` with `[0, 2, 0, 0]`

3. **Q and K scaling paths**
   Instead of a single combined scale on the QK MatMul output, SDPA applies separate `Mul(1/sqrt(head_dim))` to Q and K before the QK MatMul. Added:
   - Q path: `["Mul", "Transpose", "Reshape", "Add", "MatMul"]` with `[0, 0, 0, 0, None]`
   - K path: `["Mul", "Reshape", "Transpose", "Reshape", "Transpose", "Reshape", "Add", "MatMul"]` with `[1, 0, 0, 0, 0, 0, 0, None]` (K^T uses a `Reshape→Transpose(0,2,1)→Reshape` chain)

4. **num_heads fallback for dynamic shapes**
   SDPA models use `-1` in reshape shape tensors for dynamic dimensions, causing `get_num_heads_and_hidden_size` to return negative values. Added a fallback to user-specified `num_heads`/`hidden_size` when detected values are invalid.

### `bart_model_generator.py` (new)

Synthetic BART SDPA attention graph generator that builds a minimal but complete attention subgraph matching the SDPA topology. Tests both `with_mask=True` (decoder self-attention) and `with_mask=False` (encoder attention) variants.

### `test_attention_fusion.py`

Added `test_bart_attention_sdpa_fusion` that verifies:
- 1 Attention node is produced for each mask variant
- Correct `num_heads` attribute
- Correct `unidirectional` attribute (1 for decoder self-attention with mask, 0 for encoder)

## Test Plan
- [x] `python -m pytest test_attention_fusion.py -v` — all 10 tests pass
- [x] `lintrunner` on all 3 changed files — no issues
- [x] Verified on real exported BART SDPA model (`hf-internal-testing/tiny-random-bart`): 2 Attention nodes fused, graph reduced from 120 → 34 nodes

### CODE REVIEW

File: onnxruntime/python/tools/transformers/fusion_bart_attention.py [hunk 1]
```diff
@@ -291,19 +343,21 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
...
-        encoder_attention = one_root_input and qk_nodes == qk_nodes_no_mask
-        decoder_self_attention = one_root_input and qk_nodes == qk_nodes_with_mask
-        decoder_cross_attention = two_root_inputs and qk_nodes == qk_nodes_no_mask
+        # 1) Encoder attention with one_root_input=True and no mask
+        # 2) Decoder self attention with one_root_input=True and has mask
+        # 3) Decoder cross attention with two_root_inputs=True and no mask
+        # 4) Decoder self attention with past with one_root_input=True and has mask and past_k and past_v
+        # 5) Decoder cross attention with past with three_root_inputs=True and no mask
+        has_mask = qk_nodes in (qk_nodes_with_mask, qk_nodes_sdpa_with_mask)
+        no_mask = not has_mask
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_bart_attention.py:355]: The logic is sound, but consider documenting why this approach was chosen over a more explicit check.

File: onnxruntime/python/tools/transformers/fusion_bart_attention.py [hunk 2]
```diff
@@ -349,6 +403,12 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
             attention_last_node = reshape_qkv
             num_heads, hidden_size = self.get_num_heads_and_hidden_size(reshape_q)
 
+            # Fall back to user-specified values when detected values are invalid
+            # (e.g., SDPA models use -1 in reshape shapes for dynamic dimensions).
+            if (num_heads <= 0 or hidden_size <= 0) and self.num_heads > 0 and self.hidden_size > 0:
+                num_heads = self.num_heads
+                hidden_size = self.hidden_size
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_bart_attention.py:421]: When num_heads/hidden_size fallback is used, there's no logging to indicate this happened. Consider adding a debug log.

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/python/tools/transformers/fusion_bart_attention.py (hunk 1 above)
Comment (Rishi-Dave) [reply onnxruntime/python/tools/transformers/fusion_bart_attention.py:355]: Good point — added a comment explaining that we reuse the result of match_parent_paths above rather than re-walking the graph. Pushed in 886a566.

File: onnxruntime/python/tools/transformers/fusion_bart_attention.py (hunk 2 above)
Comment (Rishi-Dave) [reply onnxruntime/python/tools/transformers/fusion_bart_attention.py:421]: Added a logger.debug that logs both the detected (invalid) values and the user-specified fallback values. Pushed in 886a566.

**Review by tianleiwu — APPROVED** [review]
